### PR TITLE
Token binding for Oauth2.0

### PR DESCRIPTION
**Pull request by inthirakumaaran**
### Proposed changes in this pull request

Implementation of Token Binding for Oauth2.0

-

### Follow up actions

Code Review
trigger a migration in DB. Need to add new column tB_MANDATORY in IDN_OAUTH_CONSUMER_APPS

-

#### General

- [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description.
- [ ] **Is the PR labeled correctly?**

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [ ] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled.

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests.
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/).

#### Tests

- [ ] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components.
- [ ] **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- [ ] **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

- [ ] **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
- [ ] **Are all UI and API inputs run through forms or serializers?**
- [ ] **Are all external inputs validated and sanitized appropriately?**
- [ ] **Does all branching logic have a default case?**
- [ ] **Does this solution handle outliers and edge cases gracefully?**
- [ ] **Are all external communications secured and restricted to SSL?**

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes should be documented.
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented.
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

### CODE REVIEW

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java [hunk 1]
```diff
@@ -860,26 +862,35 @@ private String handleOAuthAuthorizationRequest(String clientId, HttpServletReque
             //Check if the code challenge method value is neither "plain" or "s256", if so return error
             if (pkceChallengeCode != null && pkceChallengeMethod != null) {
                 if (!OAuthConstants.OAUTH_PKCE_PLAIN_CHALLENGE.equals(pkceChallengeMethod) &&
-                        !OAuthConstants.OAUTH_PKCE_S256_CHALLENGE.equals(pkceChallengeMethod)) {
+                        !OAuthConstants.OAUTH_PKCE_S256_CHALLENGE.equals(pkceChallengeMethod)&&
```
Comment (inthirakumaaran) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java]: create a new private method to check the validity

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java [hunk 2]
```diff
@@ -860,26 +862,35 @@ private String handleOAuthAuthorizationRequest(String clientId, HttpServletReque
...
-            if (pkceChallengeCode != null && !OAuth2Util.validatePKCECodeChallenge(pkceChallengeCode, pkceChallengeMethod)) {
+            if (pkceChallengeCode != null && !OAuthConstants.OAUTH_PKCE_REFERREDTB_CHALLENGE.equals(pkceChallengeCode)
+                    && !OAuth2Util.validatePKCECodeChallenge(pkceChallengeCode, pkceChallengeMethod)) {
                 return EndpointUtil.getErrorPageURL(OAuth2ErrorCodes.INVALID_REQUEST, "Code challenge used is not up to " +
                                 "RFC 7636 specifications."
                         , null);
             }
+            log.debug("the pkce is :"+OAuthConstants.OAUTH_PKCE_REFERREDTB_CHALLENGE);
+            //check PKCE challenge for token binding
+            if (checkTB(oauthRequest.getHttpRequestHeaders()) && !OAuthConstants.OAUTH_PKCE_REFERREDTB_CHALLENGE.equals(pkceChallengeCode)){
```
Comment (inthirakumaaran) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java]: checkTokenBindingHeader  change the name

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java [hunk 3]
```diff
@@ -1385,6 +1397,17 @@ private String manageOIDCSessionState(HttpServletRequest request, HttpServletRes
 
         return redirectURL;
     }
+    private Boolean checkTB(HttpRequestHeader[] HttpRequestHeaders) {
```
Comment (inthirakumaaran) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java]: change to primitive

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/introspection/IntrospectionResponseBuilder.java [hunk 4]
```diff
@@ -198,4 +199,17 @@ public IntrospectionResponseBuilder setErrorDescription(String description) {
         parameters.put(IntrospectionResponse.Error.ERROR_DESCRIPTION, description);
         return this;
     }
+
+    public IntrospectionResponseBuilder settbh(String tbh) {
```
Comment (inthirakumaaran) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/introspection/IntrospectionResponseBuilder.java]: change the name setTbh

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/introspection/OAuth2IntrospectionEndpoint.java [hunk 5]
```diff
@@ -101,6 +101,7 @@ public Response introspect(@FormParam("token") String token, @FormParam("token_t
                 .setScope(introspectionResponse.getScope())
                 .setUsername(introspectionResponse.getUsername())
                 .setTokenType(DEFAULT_TOKEN_TYPE)
+                .settbh(introspectionResponse.getTbh())
```
Comment (inthirakumaaran) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/introspection/OAuth2IntrospectionEndpoint.java]: settbh ->setTokenBindingHash

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dao/SQLQueries.java [hunk 6]
```diff
@@ -39,7 +39,7 @@
     public static final String VALIDATE_AUTHZ_CODE_WITH_PKCE = "SELECT AUTHZ_USER, USER_DOMAIN, TENANT_ID, SCOPE, " +
             "CALLBACK_URL, TIME_CREATED,VALIDITY_PERIOD, STATE, TOKEN_ID, AUTHORIZATION_CODE, CODE_ID, SUBJECT_IDENTIFIER, " +
             "PKCE_CODE_CHALLENGE, PKCE_CODE_CHALLENGE_METHOD FROM IDN_OAUTH2_AUTHORIZATION_CODE WHERE " +
-            "CONSUMER_KEY_ID = (SELECT ID FROM IDN_OAUTH_CONSUMER_APPS WHERE CONSUMER_KEY = ?) AND AUTHORIZATION_CODE = ?";
+            "CONSUMER_KEY_ID = (SELECT ID FROM IDN_OAUTH_CONSUMER_APPS WHERE CONSUMER_KEY = ?) AND AUTHORIZATION_CODE = ? ORDER BY TIME_CREATED DESC";
```
Comment (inthirakumaaran) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dao/SQLQueries.java]: revert this change

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dao/SQLQueries.java [hunk 7]
```diff
@@ -54,24 +54,24 @@
             "IDN_OAUTH2_ACCESS_TOKEN WHERE TOKEN_ID = ?";
 
     public static final String UPDATE_TOKEN_AGAINST_AUTHZ_CODE = "UPDATE IDN_OAUTH2_AUTHORIZATION_CODE SET " +
-                                                                 "TOKEN_ID=? WHERE AUTHORIZATION_CODE=?";
+            "TOKEN_ID=? WHERE AUTHORIZATION_CODE=?";
```
Comment (inthirakumaaran) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dao/SQLQueries.java]: revert formatting

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dto/OAuth2IntrospectionResponseDTO.java [hunk 8]
```diff
@@ -102,6 +102,19 @@
     private Map<String, Object> properties = new HashMap<String, Object>();
 
     private String error;
+    /*
```
Comment (inthirakumaaran) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dto/OAuth2IntrospectionResponseDTO.java]: put *  and change the spelling

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java [hunk 9]
```diff
@@ -860,26 +862,35 @@ private String handleOAuthAuthorizationRequest(String clientId, HttpServletReque
...
 
             // If PKCE challenge code was sent, check if the code challenge is upto specifications
-            if (pkceChallengeCode != null && !OAuth2Util.validatePKCECodeChallenge(pkceChallengeCode, pkceChallengeMethod)) {
+            if (pkceChallengeCode != null && !OAuthConstants.OAUTH_PKCE_REFERREDTB_CHALLENGE.equals(pkceChallengeCode)
+                    && !OAuth2Util.validatePKCECodeChallenge(pkceChallengeCode, pkceChallengeMethod)) {
                 return EndpointUtil.getErrorPageURL(OAuth2ErrorCodes.INVALID_REQUEST, "Code challenge used is not up to " +
                                 "RFC 7636 specifications."
                         , null);
             }
+            log.debug("the pkce is :"+OAuthConstants.OAUTH_PKCE_REFERREDTB_CHALLENGE);
```
[the remark this answers is not in the document]
Comment (inthirakumaaran) [reply components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java]: I think this debugger is not needed.If any error comes page is redirected to an error page.So shall I remove this debugger?

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dto/OAuth2IntrospectionResponseDTO.java [hunk 10]
```diff
@@ -103,6 +103,20 @@
 
     private String error;
 
+    /*
+     * tbh used for token binding
+     * it contains the hash value of the binding ID.
+     */
+    private String tbh;
+
+    public String getTokenBindingHash() {
```
Comment (inthirakumaaran) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dto/OAuth2IntrospectionResponseDTO.java]: change the names of methods

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/OauthTokenIssuerImpl.java [hunk 11]
```diff
@@ -19,34 +19,79 @@
...
+
     private OAuthIssuer oAuthIssuerImpl = OAuthServerConfiguration.getInstance()
             .getOAuthTokenGenerator();
 
     public String accessToken(OAuthTokenReqMessageContext tokReqMsgCtx) throws OAuthSystemException {
-        return oAuthIssuerImpl.accessToken();
+        String tokenBindingId = OAuth2Util.checkTokenBindingHeader(tokReqMsgCtx,
+                OAuthConstants.HTTP_TB_REFERRED_HEADER_NAME);
+        String accesstoken = oAuthIssuerImpl.accessToken();
+        if (!tokenBindingId.isEmpty()) {
```
Comment (inthirakumaaran) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/OauthTokenIssuerImpl.java]: check with stringUtils.empty

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/OauthTokenIssuerImpl.java [hunk 12]
```diff
@@ -19,34 +19,79 @@
...
+            accesstoken = OAuth2Util.hashTB(tokenBindingId) + ";" + accesstoken;
+            accesstoken = baseEncodeTB(accesstoken);
```
Comment (kasunbg) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/OauthTokenIssuerImpl.java]: Change this method as base64Encode please. You are not doing anything TB specific there.

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/OauthTokenIssuerImpl.java [hunk 13]
```diff
@@ -19,34 +19,79 @@
...
+                OAuthConstants.HTTP_TB_PROVIDED_HEADER_NAME);
+        String refreshToken = oAuthIssuerImpl.refreshToken();
+        if (!tokenBindingId.isEmpty()) {
+            refreshToken = OAuth2Util.hashTB(tokenBindingId) + ":" + refreshToken;
+            refreshToken = baseEncodeTB(refreshToken);
+        }
+        return refreshToken;
+    }
+
+    public String baseEncodeTB(String token) {
```
Comment (kasunbg) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/OauthTokenIssuerImpl.java]: this shud be private.

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/OauthTokenIssuerImpl.java [hunk 14]
```diff
@@ -19,34 +19,79 @@
...
     private OAuthIssuer oAuthIssuerImpl = OAuthServerConfiguration.getInstance()
             .getOAuthTokenGenerator();
 
     public String accessToken(OAuthTokenReqMessageContext tokReqMsgCtx) throws OAuthSystemException {
-        return oAuthIssuerImpl.accessToken();
+        String tokenBindingId = OAuth2Util.checkTokenBindingHeader(tokReqMsgCtx,
+                OAuthConstants.HTTP_TB_REFERRED_HEADER_NAME);
+        String accesstoken = oAuthIssuerImpl.accessToken();
+        if (!tokenBindingId.isEmpty()) {
+            accesstoken = OAuth2Util.hashTB(tokenBindingId) + ";" + accesstoken;
```
Comment (inthirakumaaran) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/OauthTokenIssuerImpl.java]: make it configurable or  introduce better  symbol than ;

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/OauthTokenIssuerImpl.java [hunk 15]
```diff
@@ -19,34 +19,79 @@
...
+        }
+        return accesstoken;
     }
 
     public String refreshToken(OAuthTokenReqMessageContext tokReqMsgCtx) throws OAuthSystemException {
-        return oAuthIssuerImpl.refreshToken();
+        String tokenBindingId = OAuth2Util.checkTokenBindingHeader(tokReqMsgCtx,
+                OAuthConstants.HTTP_TB_PROVIDED_HEADER_NAME);
+        String refreshtoken = oAuthIssuerImpl.refreshToken();
+        if (!tokenBindingId.isEmpty()) {
```
Comment (inthirakumaaran) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/OauthTokenIssuerImpl.java]: write a private function

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/RefreshGrantHandler.java [hunk 16]
```diff
@@ -19,7 +19,9 @@
 package org.wso2.carbon.identity.oauth2.token.handlers.grant;
 
 
+import org.apache.axiom.util.base64.Base64Utils;
```
Comment (inthirakumaaran) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/RefreshGrantHandler.java]: change the import and  use util

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/OauthTokenIssuerImpl.java [hunk 17]
```diff
@@ -19,34 +19,79 @@
...
+            refreshtoken = OAuth2Util.hashTB(tokenBindingId) + ":" + refreshtoken;
```
Comment (kasunbg) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/OauthTokenIssuerImpl.java]: introduce a constant and use a better delimiter that may not lead to conflicts.

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/RefreshGrantHandler.java [hunk 18]
```diff
@@ -137,10 +140,18 @@ public boolean validateGrant(OAuthTokenReqMessageContext tokReqMsgCtx)
                 return false;
             }
 
-            if (log.isDebugEnabled()) {
-                log.debug("Refresh token validation successful for Client id : " + tokenReqDTO.getClientId() +
-                          ", Authorized User : " + validationDataDO.getAuthorizedUser() +
-                          ", Token Scope : " + OAuth2Util.buildScopeString(validationDataDO.getScope()));
+        }
+        if (!tokenBindingId.isEmpty()) {
+            String checkTokenValue = (new String(Base64Utils.decode(refreshToken), (Charsets.UTF_8))).split(":")[0];
```
Comment (kasunbg) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/RefreshGrantHandler.java]: use constant

Comment (inthirakumaaran) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/RefreshGrantHandler.java]: create a DTO

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/RefreshGrantHandler.java [hunk 19]
```diff
@@ -137,10 +140,18 @@ public boolean validateGrant(OAuthTokenReqMessageContext tokReqMsgCtx)
...
+            ;
+            if (OAuth2Util.checkUserNameAssertionEnabled()) {
+                checkTokenValue = (new String(Base64Utils.decode(checkTokenValue), (Charsets.UTF_8))).split(":")[0];
```
Comment (inthirakumaaran) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/RefreshGrantHandler.java]: change name checkTokenvalue

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/RefreshGrantHandler.java [hunk 20]
```diff
@@ -137,10 +140,18 @@ public boolean validateGrant(OAuthTokenReqMessageContext tokReqMsgCtx)
...
+            }
+            if (!checkTokenValue.equals(OAuth2Util.hashTB(tokenBindingId))) {
+                if (log.isDebugEnabled()) {
+                    log.debug("Token Binding validation failed for refresh token");
```
Comment (inthirakumaaran) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/RefreshGrantHandler.java]: change the log method

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/util/OAuth2Util.java [hunk 21]
```diff
@@ -1890,4 +1892,78 @@ public static String addUsernameToToken(AuthenticatedUser authenticatedUser, Str
         return Base64Utils.encode(tokenStrToEncode.getBytes(Charsets.UTF_8));
     }
 
+    //token binding methods
```
Comment (inthirakumaaran) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/util/OAuth2Util.java]: add proper comments

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/util/OAuth2Util.java [hunk 22]
```diff
@@ -1890,4 +1892,78 @@ public static String addUsernameToToken(AuthenticatedUser authenticatedUser, Str
...
+                    break;
+                }
+            }
+
+        }
+        return tokenBindingId;
+    }
+
+    //Hash the token binding ID in SHA256
+    public static String hashTB(String tBID) {
```
Comment (inthirakumaaran) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/util/OAuth2Util.java]: create java doc

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/util/OAuth2Util.java [hunk 23]
```diff
@@ -1890,4 +1892,78 @@ public static String addUsernameToToken(AuthenticatedUser authenticatedUser, Str
...
+            String authorizationCode = decodeSplitbase64TB(oauthCode);
+            if (!authorizationCode.equals(hashTB(codeVerifier))) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    //check for TokenBinding header in request headers
+    public static String checkTokenBindingHeader(OAuthTokenReqMessageContext tokReqMsgCtx, String httpTBheader) {
```
Comment (inthirakumaaran) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/util/OAuth2Util.java]: change name to find token binding Header value

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/util/OAuth2Util.java [hunk 24]
```diff
@@ -1890,4 +1892,78 @@ public static String addUsernameToToken(AuthenticatedUser authenticatedUser, Str
...
+        HttpRequestHeader[] httpRequestHeaders = tokReqMsgCtx.getOauth2AccessTokenReqDTO().getHttpRequestHeaders();
```
Comment (inthirakumaaran) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/util/OAuth2Util.java]: create a private method

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/util/OAuth2Util.java [hunk 25]
```diff
@@ -1890,4 +1892,78 @@ public static String addUsernameToToken(AuthenticatedUser authenticatedUser, Str
...
+        try {
+            MessageDigest messageDigest = MessageDigest.getInstance("SHA-256");
+            byte[] hash = messageDigest.digest(tBID.getBytes(StandardCharsets.US_ASCII));
+            //Trim the base64 string to remove trailing CR LF characters.
+            bindValue = new String(Base64.encodeBase64URLSafe(hash),
+                    StandardCharsets.UTF_8).trim();
+
+        } catch (NoSuchAlgorithmException e) {
+            if (log.isDebugEnabled()) {
+                log.debug("Failed to create SHA256 Message Digest.");
```
Comment (inthirakumaaran) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/util/OAuth2Util.java]: create an error log

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/validators/TokenValidationHandler.java [hunk 26]
```diff
@@ -460,4 +464,31 @@ private AccessTokenDO findAccessToken(String tokenIdentifier) throws IdentityOAu
         return OAuth2Util.getAccessTokenDOfromTokenIdentifier(tokenIdentifier);
     }
 
+    //get the hash value of token binding id
+    private String findTokenBindingHash(AccessTokenDO accessTokenDO) {
```
Comment (inthirakumaaran) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/validators/TokenValidationHandler.java]: move most of the logic to DTO

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java [hunk 27]
```diff
@@ -849,39 +851,41 @@ private String handleOAuthAuthorizationRequest(String clientId, HttpServletReque
...
+                            "This PKCE is not suitable for Token binding. ", null);
                 }
             }
-            //Check if the code challenge method value is neither "plain" or "s256", if so return error
-            if (pkceChallengeCode != null && pkceChallengeMethod != null) {
-                if (!OAuthConstants.OAUTH_PKCE_PLAIN_CHALLENGE.equals(pkceChallengeMethod) &&
-                        !OAuthConstants.OAUTH_PKCE_S256_CHALLENGE.equals(pkceChallengeMethod)) {
-                    return EndpointUtil.getErrorPageURL(OAuth2ErrorCodes.INVALID_REQUEST, "Unsupported PKCE Challenge Method"
-                            , null);
+            else {
```
Comment (inthirakumaaran) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java]: formatting

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/OauthTokenIssuerImpl.java (hunk 12 above)
Comment (inthirakumaaran) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/OauthTokenIssuerImpl.java]: okay

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/OauthTokenIssuerImpl.java (hunk 17 above)
Comment (inthirakumaaran) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/OauthTokenIssuerImpl.java]: okay

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/OauthTokenIssuerImpl.java (hunk 13 above)
Comment (inthirakumaaran) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/OauthTokenIssuerImpl.java]: okay

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/RefreshGrantHandler.java (hunk 18 above)
Comment (inthirakumaaran) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/RefreshGrantHandler.java]: okay
